controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -275,6 +282,205 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
+     * @param subscriber  Stream Subscriber publishing the StreamCut.
+     * @param truncationStreamCut  Truncation StreamCut.
+     * @param contextOpt optional context
+     * @return update status.
+     */
+    public CompletableFuture<UpdateSubscriberStatus.Status> updateSubscriberStreamCut(String scope, String stream,
+                                                                             String subscriber, ImmutableMap<Long, Long> truncationStreamCut,
+                                                                             OperationContext contextOpt) {
+        final OperationContext context = contextOpt == null ? streamMetadataStore.createContext(scope, stream) : contextOpt;
+        final long requestId = requestTracker.getRequestIdFor("updateSubscriberStreamCut", scope, stream);
+
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.checkStreamExists(scope, stream)
+                .thenCompose(exists -> {
+                            // 1. check Stream exists
+                            if (!exists) {
+                                return CompletableFuture.completedFuture(UpdateSubscriberStatus.Status.STREAM_NOT_FOUND);
+                            }
+                            // 2. check if StreamCut is valid
+                            return Futures.exceptionallyExpecting(streamMetadataStore.getSubscriber(scope, stream, subscriber, contextOpt, executor),
+                                    e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, null)
+                                    .thenCompose(subscriberRecord -> {
+                                        if (subscriberRecord == null) {
+                                            return CompletableFuture.completedFuture(UpdateSubscriberStatus.Status.SUBSCRIBER_NOT_FOUND);
+                                        } else {
+                                            if (subscriberRecord.getObject().getTruncationStreamCut().isEmpty()) {
+                                                return streamMetadataStore.isStreamCutValid(scope, stream, truncationStreamCut, context, executor)
+                                                        .thenCompose( isValid -> {
+                                                         if (isValid) {
+                                                           return updateStreamCut(scope, stream, subscriber, truncationStreamCut, context, requestId);
+                                                         } else {
+                                                           return CompletableFuture.completedFuture(UpdateSubscriberStatus.Status.STREAMCUT_NOT_VALID);
+                                                         }
+                                                 });
+                                            } else {
+                                                return streamMetadataStore.isStreamCutValidForTruncation(scope, stream, truncationStreamCut,
```

Comment:
here you are fetching subscriber record, checking its validity and then updating the subscriber record but the update is not predicated on previous subscriber record that you had fetched. this could mean if the  subscriber record was updated while you performed the check, it could get overwritten incorrectly.
this check should be done in the storage layer where you do a get followed by a check and then conditional update using the entity tag from the previous record.

Reply:
I;m fetching the subscriberRecord again in the StorageLayer and only then updating the streamcut on top. But I agree  this double fetch can be avoided.
Will change this to do the validity check and get/update directly in storage layer as done in case of RetentionSet

Reply:
Fixed.